aws-lc-rs/examples/cipher.rs

```diff
@@ -0,0 +1,233 @@
+//! ciphertext: 88bcbd8d1656d60de739c5
+//!
+//! $ cargo run --example cipher -- --mode ctr --key b331133eb742497c67ced9520c9a7de3 decrypt --iv 4e967c7b799e0670431888e2e959e154 88bcbd8d1656d60de739c5
+//! Hello World
+//!
+//! $ cargo run --example cipher -- --mode cbc encrypt "Hello World"
+//! key: 6489d8ce0c4facf18b872705a05d5ee4
+//! iv: 5cd56fb752830ec2459889226c5431bd
+//! ciphertext: 6311c14e8104730be124ce1e57e51fe3
+//!
+//! $ cargo run --example cipher -- --mode cbc --key 6489d8ce0c4facf18b872705a05d5ee4 decrypt --iv 5cd56fb752830ec2459889226c5431bd 6311c14e8104730be124ce1e57e51fe3
+//! Hello World
+//! ```
+use aws_lc_rs::{
+    cipher::{
+        CipherContext, DecryptingKey, EncryptingKey, PaddedBlockDecryptingKey,
+        PaddedBlockEncryptingKey, UnboundCipherKey, AES_128, AES_256,
+    },
+    iv::FixedLength,
+};
+use clap::{Parser, Subcommand, ValueEnum};
+
+#[derive(Parser)]
+#[command(author, version, name = "cipher")]
+struct Cli {
+    #[arg(
+        short,
+        long,
+        help = "AES 128 or 256 bit key in hex, if not provided defaults to 128"
+    )]
+    key: Option<String>,
+
+    #[arg(short, long, value_enum, help = "AES cipher mode")]
+    mode: Mode,
+
+    #[command(subcommand)]
+    command: Commands,
+}
+
+#[derive(ValueEnum, Clone, Copy)]
+enum Mode {
+    Ctr,
+    Cbc,
+}
+
+#[derive(Subcommand)]
+enum Commands {
+    Encrypt {
+        #[arg(short, long, help = "Initalization Vector (IV)")]
+        iv: Option<String>,
+        plaintext: String,
+    },
+    Decrypt {
+        #[arg(short, long, help = "Initalization Vector (IV)")]
+        iv: String,
+        ciphertext: String,
+    },
+}
+
+fn main() -> Result<(), &'static str> {
+    let cli = Cli::parse();
+
+    let key = if let Some(key) = cli.key {
+        match hex::decode(key) {
+            Ok(v) => v,
+            Err(..) => {
+                return Err("invalid key");
+            }
+        }
+    } else {
+        let mut v = vec![0u8; 16];
+        aws_lc_rs::rand::fill(v.as_mut_slice()).map_err(|_| "failed to generate key")?;
+        v
+    };
+
+    match (cli.command, cli.mode) {
+        (Commands::Encrypt { iv, plaintext }, Mode::Ctr) => aes_ctr_encrypt(&key, iv, plaintext),
+        (Commands::Encrypt { iv, plaintext }, Mode::Cbc) => aes_cbc_encrypt(&key, iv, plaintext),
+        (Commands::Decrypt { iv, ciphertext }, Mode::Ctr) => aes_ctr_decrypt(&key, iv, ciphertext),
+        (Commands::Decrypt { iv, ciphertext }, Mode::Cbc) => aes_cbc_decrypt(&key, iv, ciphertext),
+    }?;
+
+    Ok(())
+}
+
+fn aes_ctr_encrypt(key: &[u8], iv: Option<String>, plaintext: String) -> Result<(), &'static str> {
+    let hex_key = hex::encode(key);
+    let key = new_unbound_key(key)?;
+
+    let key = match iv {
+        Some(iv) => {
+            let iv = {
+                let v = hex::decode(iv).map_err(|_| "invalid iv")?;
+                let v: FixedLength<16> = v.as_slice().try_into().map_err(|_| "invalid iv")?;
```

Comment:
Can we use `AES_IV_LEN` here? Same for the other instances below, it might be best to show in our examples that the user can depend on the predefined values we provide.
```suggestion
                let v: FixedLength<AES_IV_LEN> = v.as_slice().try_into().map_err(|_| "invalid iv")?;
```